src/cycles/eulerian.jl

```diff
@@ -0,0 +1,315 @@
+# Checking for existence of Eulerian circuit and Eulerian trail
+
+function single_nonzero_degree_component(
+    graph::AG,
+    ) where {T <: Integer, AG <: AbstractGraph{T}}
+
+    visited = zeros(UInt8, nv(graph))
+    for v in vertices(graph)
+        if (outdegree(graph, v) + indegree(graph, v)) == 0
+            visited[v] = 2
+        end
+    end
+    for v in vertices(graph)
+        if (outdegree(graph, v) + indegree(graph, v)) == 0
+            continue
+        end
+        stack = Vector{T}([v])
+        visited[v] = 1
+        while !isempty(stack)
+            u = stack[end]
+            # assuming valid vertices never take zero value
+            w = zero(eltype(graph))
+            for v in outneighbors(graph, u)
+                if visited[v] == 0
+                    w = v
+                    break
+                end
+            end
+            if w != zero(eltype(graph))
+                visited[w] = 1
+                push!(stack, w)
+            else
+                visited[u] = 2
+                pop!(stack)
+            end
+        end
+        break
+    end
+    for v in vertices(graph)
+        visited[v] > 0 || return false
+    end
+    return true
+end
+
+"""
+    has_eulerian_trail(g)
+
+Return `true` if graph `g` contains [Eulerian trail](https://en.wikipedia.org/wiki/Eulerian_path).
+Note that a Eulerian Trail can have the same start and end vertex so all Eulerian Circuits are Trails as well.
```

Comment:
Is there a reason, that we also allow eulerian circuits to pass this function?

Reply:
As initially, the issue was just to check if eulerian trails exist or not so I kept this if anyone just wants to check but yeah, this can also be checked if ```eulerian_trail() ``` throws ArgumentError.

Reply:
It is kind of strange, because we also have the function `has_eulerian_circuit, so I think `has_eulerian_trail` should only test for trails.

Reply:
But like in case of a cyclic graph it is wrong to say that it has no eulerian trail since every eulerian circuit is a eulerian trail. Also the code is not exclusively checking for circuits, from the comments I just wanted to say that the returned trail may also be a circuit for a given graph.

Reply:
I think the issue here might be, that we have slightly different definitions of what an eulerian trail is. I think, that a trail cannot share the start and end vertex, so for example a graph that is a simple cycle, has an eulerian circuit but not a trail, because one cannot traverse all edges in this graph without having the same start and end vertex.

Reply:
I referred Wikipedia for the definition https://en.wikipedia.org/wiki/Eulerian_path and on the first para, they say "a Eulerian circuit or Eulerian cycle is a Eulerian trail which starts and ends on the same vertex."

Reply:
You are totally right, I had read that incorrectly on Wikipeda.
This commes a bit late in this PR, but would you agree with me, that it is not very useful to have an algorithm that covers the case of open and closed trails, when the open case is already covered?

I wonder if it would better (from a user perspective), if we had an algorithm, that decides for two starting vertices x, y (not necessarily distinct), if there is an eulerian trail from x to y and an algorithm that returns this trail.

Then we could leave the `eulerian_circuit` and `has_eulerian_circuit` functions untouched and would change the functions for the trail.

What do you think?

Reply:
Do you mean closed case already covered by "when the open case is already covered?" in `eulerian_circuit()`?
Yes, I totally agree it's not very useful to just have a function that returns any trail in the graph. Having the desired source-destination would be really nice.

Reply:
Yes exactly

Reply:
I will see the algorithm again and will try to change it.

Reply:
I think I can do this: Since trail will exist in two cases zero odd degree nodes or two odd degree nodes (assuming a connected graph for simple discussion).
In case of no odd degree nodes a trail will always be a circuit so trail(u, v) will return no trail for u != v and a circuit for u == v.
In case of two odd degree nodes: For an undirected graph, a trail will be between the two odd degree nodes, so if deg(u) or deg(v) is even no trail exists between them, otherwise return one of the existing trails between u and v.
For directed graph, trails will only exist from a node with out-in == 1 to in-out == 1 node. so if u,v do not satisfy this then no trail otherwise return one of the existing trails between u and v.

Reply:
I have done some changes and now `eulerian_trail()` is the only function that will return trail/circuit between asked source-destination pair. `eulerian_circuit()` and `has_eulerian_circuit()` are useless now and `has_eulerian_trail()` is just a helper function so it is not exported.
Here is `eulerian_trail()` for directed graphs
```
function eulerian_trail end
@traitfn function eulerian_trail(
    graph::AG::(IsDirected),
    snode::T,
    dnode::T
    ) where {T, AG <: AbstractGraph{T}}
    
    odd_node = [false]
    has_eulerian_trail(graph, odd_node) || throw(ArgumentError("Graph do not have Eulerian trail."))
    trail = Vector{T}()
    iszero(nv(graph)) && return trail
    iszero(ne(graph)) && return [vertices(graph)[1]]
    
    if ((odd_node[1] == false) && (snode != dnode)) || ((odd_node[1] == true) && (snode == dnode))
        throw(ArgumentError("Graph do not have Eulerian trail."))
    # if eulerian trail is present but not between required source-destination pair
    elseif (odd_node[1] == true) && (snode != dnode)
        if (outdegree(graph, snode) - indegree(graph, snode) != 1) || (indegree(graph, dnode) - outdegree(graph, dnode) != 1)
            throw(ArgumentError("Graph do not have Eulerian trail."))
        end
    end

    edge_count = collect(outdegree(graph))
    current_vertex = snode
    current_path = Vector{T}([current_vertex])

    while !isempty(current_path)
        if !iszero(edge_count[current_vertex])
            push!(current_path, current_vertex)
            next_vertex = neighbors(graph, current_vertex)[edge_count[current_vertex]]
            edge_count[current_vertex] -= 1
            current_vertex = next_vertex
        else
            push!(trail, current_vertex)
            current_vertex = current_path[end]
            pop!(current_path)
        end
    end
    
    return reverse!(trail)
end
```
I have not pushed the changes because these require changes unit tests according to the desired function type but I have checked this on some cases. So I thought to first discuss the main function then change unit test accordingly.